pkg/config/mesh/kubemesh/watcher.go

```diff
@@ -18,112 +18,50 @@ import (
 	"fmt"
 
 	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
 
-	meshconfig "istio.io/api/mesh/v1alpha1"
-	"istio.io/istio/pkg/config/mesh"
+	"istio.io/istio/pkg/config/mesh/meshwatcher"
 	"istio.io/istio/pkg/kube"
-	"istio.io/istio/pkg/kube/watcher/configmapwatcher"
-	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/kube/kclient"
+	"istio.io/istio/pkg/kube/krt"
+	"istio.io/istio/pkg/ptr"
 )
 
-// NewConfigMapWatcher creates a new Watcher for changes to the given ConfigMap.
-func NewConfigMapWatcher(client kube.Client, namespace, name, key string, multiWatch bool, stop <-chan struct{}) *mesh.MultiWatcher {
-	w := mesh.NewMultiWatcher(mesh.DefaultMeshConfig())
-	c := configmapwatcher.NewController(client, namespace, name, func(cm *v1.ConfigMap) {
-		meshNetworks, err := ReadNetworksConfigMap(cm, "meshNetworks")
-		if err != nil {
-			// Keep the last known config in case there's a misconfiguration issue.
-			log.Errorf("failed to read meshNetworks config from ConfigMap: %v", err)
-			return
-		}
-		if meshNetworks != nil {
-			w.SetNetworks(meshNetworks)
-		}
-		if multiWatch {
-			meshConfig := meshConfigMapData(cm, key)
-			w.HandleMeshConfigData(meshConfig)
-			return
-		}
-		// Original behavior - just per-revision config
-		meshConfig, err := ReadConfigMap(cm, key)
-		if err != nil {
-			// Keep the last known config in case there's a misconfiguration issue.
-			log.Errorf("failed to read mesh config from ConfigMap: %v", err)
-			return
-		}
-		w.HandleMeshConfig(meshConfig)
-	})
-
-	go c.Run(stop)
-
-	// Ensure the ConfigMap is initially loaded if present.
-	if !client.WaitForCacheSync("configmap watcher", stop, c.HasSynced) {
-		log.Error("failed to wait for cache sync")
-	}
-	return w
-}
+const (
+	MeshConfigKey   = "mesh"
+	MeshNetworksKey = "meshNetworks"
+)
 
-func AddUserMeshConfig(client kube.Client, watcher mesh.Watcher, namespace, key, userMeshConfig string, stop <-chan struct{}) {
-	c := configmapwatcher.NewController(client, namespace, userMeshConfig, func(cm *v1.ConfigMap) {
-		meshConfig := meshConfigMapData(cm, key)
-		watcher.HandleUserMeshConfig(meshConfig)
+func NewConfigMapSource(client kube.Client, namespace, name, key string, opts krt.OptionsBuilder) meshwatcher.MeshConfigSource {
+	clt := kclient.NewFiltered[*v1.ConfigMap](client, kclient.Filter{
+		Namespace:     namespace,
+		FieldSelector: fields.OneTermEqualSelector(metav1.ObjectNameField, name).String(),
 	})
-
-	go c.Run(stop)
-	if !client.WaitForCacheSync("user mesh config", stop, c.HasSynced) {
-		log.Error("failed to wait for cache sync")
-	}
-}
-
-func meshConfigMapData(cm *v1.ConfigMap, key string) string {
-	if cm == nil {
-		return ""
-	}
-
-	cfgYaml, exists := cm.Data[key]
-	if !exists {
-		return ""
-	}
-
-	return cfgYaml
+	cms := krt.WrapClient(clt, opts.WithName("ConfigMap_"+name)...)
```

Comment:
nit: would be nice if this was only set if the opts didn't already contain a name for the source. That way someone could have a custom name specifically for revisions or something (assuming this is more generic than just meshconfig)

Reply:
`krt.OptionsBuilder` doesn't have the ability to have a name currently so that couldn't happen. If we have a use case, I would imagine we could add that capability though